WT-6010 Workgen changes - Add session config at Thread level.

raviprakashgiri29 opened the pull request:
Looks like some Operation flow in workgen.cxx calls cursor operations without calling begin_transaction().

Updated workgen.cxx to call any cursor operation only after begin_transaction() (_in_transaction = true).

ddanderson commented on bench/workgen/workgen.cxx:
```diff
@@ -882,7 +882,8 @@ int ThreadRunner::op_run(Operation *op) {
             _in_transaction = true;
         }
-        if (op->is_table_op()) {
+        // Call cursor operations only when transaction != NULL or _in_transaction is true.
+        if (op->is_table_op() && _in_transaction) {
```
@raviprakashgiri29, I don't think this is right.  Doesn't this say that if you're not using transactions, no operations occur?  We want to support workloads that have no transactions.

ddanderson replied:
I suppose we could have logic that said if some transactions used one isolation level and other transactions used another one, or didn't explicitly use a transaction, then we would issue some sort of error within workgen.  But that's a lot more work than we want to do to catch a script error.  I think just fixing the script is the right solution.

raviprakashgiri29 replied:
@ddanderson : I don't see any issue in the script. UPDATE operation in the script is wrapped under a transaction (`txn`) with snapshot isolation level but the code flow hits the UPDATE `cursor->update()` without snapshot isolation level. 
I think some issue in the workgen.cxx file in this case wherein the code calls update operation `cursor->update()` without setting `begin_transaction isolation level` to snapshot and by default the snapshot level is `read-committed`.

ddanderson replied:
Sorry if I wasn't clear.  I think your changes to `example_prepare.py` are fine.  I don't think the changes to `workgen.cxx` are correct, as that will have the effect of having scripts that don't use transactions (like `example_simple.py`) not do any work (no inserts will actually be done).  To make sure I applied just the `example_prepare.py` changes, and running that example worked.  Let's check in on this next week.

raviprakashgiri29 replied:
@ddanderson : Thank you for correcting me. I understand your concern here and I agree that checking for `_in_transaction` will block workloads running without any transaction. 

I will revert the change and update the PR to pass the transaction begin_config i.e, into `conn->open_session()`. 
ex: workloads running without transaction will pass the NULL config string and it is backward compatible.

kommiharibabu commented:
@raviprakashgiri29  The PR title doesn't represent the changes that are done, please update the PR title before merging it according to the changes.

raviprakashgiri29 commented:
> @raviprakashgiri29, there's some parts about this that I don't understand. The script puts some configuration on the transaction, but that is translated in `workgen.cxx` to configuration on the session. Second, the set of thread operations can be very complex. While it can be a single operation or transaction, it can also be a group (for example, a sleep followed by a transaction). So looking at the `_thread->_op` is looking at the first operation (or the top operation, if it's a group). If we need to put configuration on certain sessions, then I expect we need a new config variable in the `ThreadOptions` class. Maybe `session_config`? Then the Python script would set `thread.options.session_config = ...`. Let's talk about this.

@ddanderson : I agree that transaction can either be a single Operation or set of Operations and using transaction-level configurations for session config is not ideal. idea. I have updated the `ThreadOptions` class and added a new session configuration variable on the thread level which would be passed to `open_session()`.

ddanderson commented on bench/workgen/runner/example_prepare.py:
```diff
@@ -85,6 +89,6 @@
 
 print('Workload took %d minutes' %(run_time//60))
 
-latency_filename = homedir + "/latency.out"
+latency_filename = "WT_TEST" + "/latency.out"
```
I think I prefer keeping the variable `homedir` here, that way the home directory can be changed in one place.

raviprakashgiri29 replied:
Fixed.

ddanderson commented on bench/workgen/workgen.h:
```diff
@@ -342,6 +343,7 @@ struct ThreadOptions {
 	os << "throttle " << throttle;
 	os << ", throttle_burst " << throttle_burst;
 	os << ", synchronized " << synchronized;
+    os << ", session_config " << session_config;
```
The spacing looks different from the previous line.

raviprakashgiri29 replied:
It was some weird spacing issue. I had to copy the above os statement and edit it.

raviprakashgiri29 commented:
> @raviprakashgiri29 The PR title doesn't represent the changes that are done, please update the PR title before merging it according to the changes.

Thank you @kommiharibabu. Fixed the title.
